File: examples_Fortran/04_eigenproblem_elpa/eigenproblem_elpa.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/time.h>
#include <math.h>
#include "mpi.h"

#include <elpa/elpa.h>
#include <assert.h>

#define assert_elpa_ok(x) assert(x == ELPA_OK)

static int max( int a, int b ){if (a>b) return(a); else return(b);}
static int min( int a, int b ){if (a<b) return(a); else return(b);}

//__________________________________________________________________________________

int N = 1000;
int nev=64;
int NB = 32; 

int debug_mode=0;

enum matrix_types {Symmetrized_Clement, Random_0_1};
enum matrix_types matrix_type = Symmetrized_Clement;

int ELPA_SOLVER = ELPA_SOLVER_2STAGE;

//__________________________________________________________________________________
// blacs
void   Cblacs_pinfo( int* mypnum, int* world_size);
void   Cblacs_get( int context, int request, int* value);
int    Cblacs_gridinit( int* context, char * order, int np_row, int np_col);
void   Cblacs_gridinfo( int context, int*  np_row, int* np_col, int*  my_row, int*  my_col);
void   Cblacs_gridexit( int context);
void   Cblacs_exit( int error_code);

// scalapack
int    numroc_( int *n, int *NB, int *iproc, int *isrcproc, int *world_size);
void   descinit_( int *desc, int *m, int *n, int *mb, int *NB, int *irsrc, int *icsrc, int *ictxt, int *lld, int *info);

//__________________________________________________________________________________


int Setup_ELPA()
	{
	return 0;	
	}

double get_global_matrix_element (int I_gl, int J_gl)
   {
   double matrix_element=0;
   
   if (matrix_type==Symmetrized_Clement)
      {
      if (I_gl==J_gl+1 || J_gl==I_gl+1) 
         {
         int K = min(I_gl, J_gl);
         matrix_element = (double) sqrtl((K+1) * (N - K-1));
         }
      }
   else if (matrix_type==Random_0_1)
      {
      matrix_element = (double) (rand())/RAND_MAX;
      }
   else
      {
      printf("matrix_type=%d is not supported", matrix_type);
      exit(1);
      }
   
   return matrix_element;
   }

//__________________________________________________________________________________
	
int main(int argc, char **argv) 
{
//____________________________________________ 
// setup MPI
int world_rank, world_size; // MPI

// for hybrid MPI+OpenMP
//int thread_level;
//MPI_Init_thread(&argc, &argv, MPI_THREAD_FUNNELED, &thread_level); 

MPI_Init( &argc, &argv);
MPI_Comm_size(MPI_COMM_WORLD, &world_size);
MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);

//____________________________________________ 

//int world_rank, world_size; // blacs
int ictxt, nprow, npcol, myrow, mycol;
int np, nq;
int m_loc, n_loc;
int info, itemp, seed, lwork;
int descA[9], descZ[9];
int izero=0;
//double *A_loc; // this is the matrix we diagonalize here
//double *Z_loc, *work, *Eigenvalues;
  


//____________________________________________ 

if (argc==1) // one argument was provided: filename (default)
   {
   N = 10;
   nev = N;
   }
   
if (argc==2) // two arguments were provided: filename (default), N
   {
   N   = atoi(argv[1]);
   nev = N;
   }

if (argc==3) // three arguments were provided: filename (default), N, nev
   {
   N   = atoi(argv[1]);
   nev = atoi(argv[2]);
   }

if (argc==4) // four arguments were provided: filename (default), N, nev, NB
   {
   N   = atoi(argv[1]);
   nev = atoi(argv[2]);
   NB  = atoi(argv[3]);
   }

//____________________________________________ 

nprow = sqrt(world_size); 
npcol = sqrt(world_size); 
if (world_size==2)  {nprow=2; npcol=1;}
if (world_size==6)  {nprow=3; npcol=2;}

if (world_size==10) {nprow=2; npcol=5;}
if (world_size==20) {nprow=4; npcol=5;}
if (world_size==40) {nprow=5; npcol=8;}

if (world_size==72) {nprow=8; npcol=9;}
//if (world_size==72) {nprow=9; npcol=8;}
//if (world_size==72) {nprow=6; npcol=12;}
//if (world_size==72) {nprow=12; npcol=6;}

if (world_size==72*4) {nprow=8*2; npcol=9*2;}

if (NB>N/max(nprow,npcol)) NB = N/max(nprow,npcol);

if (nprow*npcol!=world_size)
	{
  if (world_rank==0)
  printf("ERROR:  wrong grid \n");
  MPI_Finalize(); 
	return(1);
  }

if (debug_mode) printf("world_rank= %i \n", world_rank);
if (world_rank==0) printf("world_size=%i, nprow=%i, npcol=%i, N=%i, nev=%i, NB=%i \n", world_size, nprow, npcol, N, nev, NB);

//____________________________________________ 
	
// setup Blacs
Cblacs_pinfo( &world_rank, &world_size ) ;
Cblacs_get( -1, 0, &ictxt );
Cblacs_gridinit( &ictxt, "Row", nprow, npcol );
Cblacs_gridinfo( ictxt, &nprow, &npcol, &myrow, &mycol );

// Compute the size of the local matrices (thanks to numroc)
m_loc = numroc_( &N, &NB, &myrow, &izero, &nprow );
n_loc = numroc_( &N, &NB, &mycol, &izero, &npcol );
if (debug_mode) printf("myrow=%i, mycol=%i, m_loc=%i, n_loc=%i, \n", myrow, mycol, m_loc, n_loc);

//____________________________________________ 

/* Setup ELPA */
///*
int error, status;
elpa_t handle;
   
if (elpa_init(20170403) != ELPA_OK) 
	{
	fprintf(stderr, "Error: ELPA API version not supported");
	exit(1);
	}

handle = elpa_allocate(&error);
assert_elpa_ok(error);

//elpa_set(handle, "na", n, &error); // matrix dimension
elpa_set_integer(handle, "na", N, &error); // matrix dimension
if (world_rank==0 || debug_mode) printf("na=%i is set, error=%i\n", N, error);

elpa_set_integer(handle, "nev", nev, &error); // ?? number of eigenvalues/eigenvectors?
if (world_rank==0 || debug_mode) printf("nev=%i is set, error=%i\n", nev, error);

elpa_set_integer(handle, "local_nrows", m_loc, &error); // m_loc
if (world_rank==0 || debug_mode) printf("local_nrows=%i is set, error=%i\n", m_loc, error);

elpa_set_integer(handle, "local_ncols", n_loc, &error); // n_loc
if (world_rank==0 || debug_mode) printf("local_ncols=%i is set, error=%i\n", n_loc, error);

elpa_set_integer(handle, "nblk", NB, &error); // NB
if (world_rank==0 || debug_mode) printf("nblk=%i is set, error=%i\n", NB, error);

elpa_set_integer(handle, "mpi_comm_parent", MPI_Comm_c2f(MPI_COMM_WORLD), &error);
if (world_rank==0 || debug_mode) printf("Fortran MPI communicator is set, error=%i\n", error);

elpa_set_integer(handle, "process_row", myrow, &error); // myrow
if (world_rank==0 || debug_mode) printf("process_row=%i is set, error=%i\n", myrow, error);

elpa_set_integer(handle, "process_col", mycol, &error); // mycol
if (world_rank==0 || debug_mode) printf("process_col=%i is set, error=%i\n", mycol, error);

int error1 = elpa_setup(handle);
printf("ELPA setup done, error=%i\n", error1);
if (error1!=ELPA_OK)
  {
  printf("ELPA setup failed, error=%i\n", error1);
  MPI_Finalize();
  return(1);
  }
//if(world_rank == 0) printf("ELPA setup done\n");
// Set ELPA tunables 
elpa_set_integer(handle, "solver", ELPA_SOLVER, &error);
if (world_rank==0 || debug_mode) 
  {
  printf("elpa_set solver done, error=%d \n", error);
  if (ELPA_SOLVER == ELPA_SOLVER_1STAGE) printf("ELPA_SOLVER_1STAGE \n");
  if (ELPA_SOLVER == ELPA_SOLVER_2STAGE) printf("ELPA_SOLVER_2STAGE \n");
  }
  
MPI_Barrier(MPI_COMM_WORLD);   
// End Setup ELPA 


// parameters of the matrix


 
// Work only the process in the process grid

double *A_loc, *Z_loc, *work, *Eigenvalues;
//vector<double> A_loc; // this is the matrix we diagonalize here
//vector<double> Z_loc, work, Eigenvalues;


// Allocate the matrices A_loc, Z_loc(vector of eigenvalues), Eigenvalues
A_loc = (double *)calloc(m_loc*n_loc,sizeof(double)) ;
Z_loc = (double *)calloc(m_loc*n_loc,sizeof(double)) ;
Eigenvalues = (double *)calloc(N,sizeof(double)) ;
  

// fill the local part of matrix A 
int k = 0;
int i_loc, j_loc;
for(int i_loc=0; i_loc<m_loc; i_loc++) // iteration over the first state
    {
    int l_1 = i_loc/NB; // local coord of the (NBxNB) block among other blocks
    int x_1 = i_loc%NB; // local coord within the block
    int I_gl= (l_1*nprow + myrow)*NB + x_1; // gl-global; nprow = "P_r"; myrow="p_r"  (quoted-ScaLAPACK userguide notation, p.88-90)

    for(int j_loc=0; j_loc<n_loc; j_loc++) // iteration over the second state
        {
        int l_2 = j_loc/NB; // local coord of the (NBxNB) block among other blocks
        int x_2 = j_loc%NB; // local coord within the block
        int J_gl= (l_2*npcol + mycol)*NB + x_2;

        A_loc[i_loc+j_loc*m_loc] = get_global_matrix_element(I_gl, J_gl);
        }
    }

      
// Initialize the array descriptor for the distributed matrices xA, U and VT
itemp = max( 1, m_loc );
descinit_( descA,  &N, &N, &NB, &NB, &izero, &izero, &ictxt, &itemp, &info );
descinit_( descZ,  &N, &N, &NB, &NB, &izero, &izero, &ictxt, &itemp, &info );



double t_start = MPI_Wtime();

elpa_eigenvectors(handle, A_loc, Eigenvalues, Z_loc, &error); // Eigenvalues=eigenvector matrix, Z_loc =eigenvalue vector

double t_stop = MPI_Wtime();
	

	

if (world_rank==0)
  {
  printf("Diagonalization is done\n");
  
  int N_eigenvalues = min(10, nev);
  printf("First %d eigenvalues: \n", N_eigenvalues);
  int i; // for old Intel compilers
  for (i=0; i<N_eigenvalues; i++) printf("%g ",Eigenvalues[i]);
  printf("\n");
  
  if (matrix_type==Symmetrized_Clement)
    {
    printf("Absolute diff of eigenvalues wrt to analytic ones: \n");
    for (i=0; i<N_eigenvalues; i++) printf("%.16g ", Eigenvalues[i] - (-N+1+2*i) );
    printf("\n");
    }
  
  printf("Diagonalization time (sec): \n%g\n", t_stop-t_start);
  }

// clean up

free(A_loc);
free(Z_loc);
free(Eigenvalues);

Cblacs_gridexit(0);

elpa_deallocate(handle, &error);
elpa_uninit(&error);
		
MPI_Finalize();

return(0);
}

File: examples_Fortran/04_eigenproblem_elpa/eigenproblem_elpa.f90
program eigenproblem_elpa
  use mpi
  ! Step 1: use the ELPA module
  use elpa
  implicit none
  
  ! Explicit interface for numroc
  interface 
    integer function numroc(N, NB, iproc, isrcproc, nprocs)
      integer, intent(in) :: N, NB, iproc, isrcproc, nprocs
    end function numroc
  end interface
  
  ! Step 2: define a handle for an ELPA object
  class(elpa_t), pointer :: elpaInstance
  integer :: status

  ! Declare global variables
  integer, parameter :: N_DEFAULT = 1000, NEV_DEFAULT = 500, NB_DEFAULT = 32
  integer :: N = N_DEFAULT, nev = NEV_DEFAULT, NB = NB_DEFAULT
  integer :: debug_mode = 0
  
  integer :: world_rank, world_size
  integer :: iam, nprocs, ictxt, nprow, npcol, myrow, mycol
  integer :: np, nq, m_loc, n_loc, info, itemp, seed
  integer :: descA(9), descZ(9)
  integer :: izero = 0, ione = 1
  integer :: i_arg, argc
  character(len=32) :: argv
  real(8) :: t0, t1, t2
  real(8), allocatable :: A_loc(:,:), Z_loc(:,:), Eigenvalues(:)
  integer :: I_loc, J_loc, I_global, J_global
  integer :: lwork, liwork
  integer :: ierr
  integer :: l_1, x_1, I_gl
  integer :: l_2, x_2, J_gl
  integer :: i, N_eigenvalues_print

  ! ____________________________________________
  ! Set up MPI
  call MPI_Init(ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD, world_size, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank, ierr)
  
  ! ____________________________________________ 
  ! Process command-line arguments
  argc = command_argument_count()
  
  if (argc == 0) then
    N = 10
    nev = N
  end if
  
  if (argc == 1) then
    call get_command_argument(1, argv)
    read(argv, *) N
    nev = N
  end if
  
  if (argc == 2) then
    call get_command_argument(1, argv)
    read(argv, *) N
    call get_command_argument(2, argv)
    read(argv, *) nev
  end if
  
  if (argc >= 3) then
    call get_command_argument(1, argv)
    read(argv, *) N
    call get_command_argument(2, argv)
    read(argv, *) nev
    call get_command_argument(3, argv)
    read(argv, *) NB
  end if

  if (N > 100) debug_mode = 0

  ! ____________________________________________ 
  ! determine the blacs grid size
  nprow = nint(sqrt(real(world_size)))
  npcol = world_size / nprow
  
  ! if the grid is not square, try to find the "most square" rectangular grid
  do while (nprow > 0)
    if (mod(world_size, nprow) == 0) then
        npcol = world_size / nprow
        exit
    end if
    nprow = nprow - 1
  end do
  
  ! try some other grids if you want
  ! if (world_size==72) {nprow=12; npcol=6;}

  ! specially treat a case of small matrices
  if (NB > N/max(nprow, npcol)) NB = N/max(nprow, npcol)

  if (nprow*npcol /= world_size) then
    if (world_rank == 0) then
        print *, 'Error:  wrong grid size'
    end if
    call MPI_Finalize(ierr) 
    stop 1
  end if

  print *, "N=", N, ", nev=", nev, ", NB=", NB

  ! ____________________________________________ 
  ! Set up blacs grid
  call blacs_pinfo(iam, nprocs)
  call blacs_get(0, 0, ictxt)
  call blacs_gridinit(ictxt, 'row', nprow, npcol)
  call blacs_gridinfo(ictxt, nprow, npcol, myrow, mycol)
  
  ! Compute the size of the local matrices using numroc
  m_loc = numroc(N, NB, myrow, 0, nprow)
  n_loc = numroc(N, NB, mycol, 0, npcol)
  if (debug_mode == 0) write(*,*) 'myrow=', myrow, ', mycol=', mycol, ', m_loc=', m_loc, ', n_loc=', n_loc

  ! Initialize array descriptors for distributed matrices A and Z
  itemp = max( 1, m_loc );
  print *, "m_loc=", m_loc, ", n_loc=", n_loc, ", itemp=", itemp
  call descinit(descA, N, N, NB, NB, 0, 0, ictxt, itemp, info)
  call descinit(descZ, N, N, NB, NB, 0, 0, ictxt, itemp, info)

  ! ____________________________________________ 
  ! Setup ELPA 
  
  ! Step 3: initialize the ELPA library
  if (elpa_init(20170403) /= ELPA_OK) then
      print *, "ELPA API version not supported"
    stop 1
  endif
  
  ! Step 4: allocate the ELPA object
  elpaInstance => elpa_allocate(status)
  ! Check status code, e.g. with
  if (status /= ELPA_OK) then
    print *, "Could not allocate ELPA instance"
    stop 1
  endif

  ! Step 5: set mandatory parameters describing the matrix and its MPI distribution
  ! This can be done only once per ELPA object (handle)
  call elpaInstance%set("na", N, status)
  if (status /= ELPA_OK) then
    print *, "Could not set parameter na"
    ! Handle this error in your application
  endif
  
  call elpaInstance%set("nev", nev, status)
  ! Check status code ...
  
  call elpaInstance%set("local_nrows", m_loc, status)
  ! Check status code ...
  
  call elpaInstance%set("local_ncols", n_loc, status)
  ! Check status code ...
  
  call elpaInstance%set("nblk", NB, status)
  ! Check status code ...
  
  call elpaInstance%set("mpi_comm_parent", MPI_COMM_WORLD, status)
  ! Check status code ...
  
  call elpaInstance%set("process_row", myrow, status)
  ! Check status code ...
  
  call elpaInstance%set("process_col", mycol, status)
  ! Check status code ...

  ! Step 6: set up the elpa object, finalize setting of mandatory parameters
  status = elpaInstance%setup()
  if (status /= ELPA_OK) then
    print *, "Could not setup the ELPA object"
    ! Handle this error in your application
  endif

  ! Step 7: set runtime options, e.g. GPU settings
  call elpaInstance%set("solver", ELPA_SOLVER_2STAGE, status)
  ! Check status code ...

  ! End of Setup ELPA 
  ! ____________________________________________ 
  ! Allocate the matrices A_loc, Z_loc (matrix of eigenvectors), and vector of Eigenvalues
  allocate(A_loc(m_loc, n_loc)) ! matrix to be diagonalized, local
  allocate(Z_loc(m_loc, n_loc)) ! matrix of eigenvectors,    local
  allocate(Eigenvalues(N))     ! vector of eigenvalues,     global

  ! Fill the local matrix A_loc 
  do i_loc = 1, m_loc
    l_1 = (i_loc-1)/NB     ! local coord of the (NBxNB) block among other blocks
    x_1 = mod(i_loc-1, NB) ! local coord within the block
    I_gl = (l_1*nprow + myrow)*NB + x_1 + 1 ! gl-global; nprow = "P_r"; myrow="p_r"  (quoted-ScaLAPACK userguide notation, p.88-90)
  
    do j_loc = 1, n_loc
      l_2 = (j_loc-1)/NB
      x_2 = mod(j_loc-1, NB)
      J_gl = (l_2*npcol + mycol)*NB + x_2 + 1
  
      A_loc(i_loc, j_loc) = get_global_matrix_element(I_gl, J_gl, N)
    end do
  end do
  
  ! ____________________________________________ 
  ! print the local part of matrix A
  if (debug_mode == 1 .and. world_rank == 0) then
    write(*,*) "A_loc:"
    do i_loc = 1, m_loc
       do j_loc = 1, n_loc
          write(*,'(F6.2,X)', advance='no'), A_loc(i_loc, j_loc)
       end do
       write(*,*)
    end do
    write(*,*)
  end if

  ! ____________________________________________ 
  ! Print ELPA settings
  call elpaInstance%print_settings(status)

  ! ____________________________________________ 
  ! Step 8: Perform the diagonalization using ELPA
  
  call MPI_Barrier(MPI_COMM_WORLD, ierr) ! for timing
  t1 = MPI_Wtime()

  call elpaInstance%eigenvectors(A_loc, Eigenvalues, Z_loc, status)
    
  call MPI_Barrier(MPI_COMM_WORLD, ierr)
  t2 = MPI_Wtime()

  ! ____________________________________________ 
  ! Print the results
  if (world_rank == 0) then
    write(*,*) "Diagonalization is done"
    if (status /= ELPA_OK) write(*,*) "ELPA status=", status, " error occured!!!"

    N_eigenvalues_print = min(10, nev)
    write(*,*) "First ", N_eigenvalues_print, " eigenvalues: "
    do i = 1, N_eigenvalues_print
        write(*, '(e23.16, " ")', advance='no') Eigenvalues(i)
    end do
    write(*,*)

    !if (matrix_type == Symmetrized_Clement) then
        write(*,*) "Absolute diff of eigenvalues wrt to analytic ones: "
        do i = 1, N_eigenvalues_print
            write(*, '(e23.16, " ")', advance='no') Eigenvalues(i) - (-N + 1 + 2*(i-1))
        end do
        write(*,*)
    !end if

    write(*,*) "Diagonalization time  (sec): "
    write(*,*) t2 - t1
  end if

  ! ____________________________________________ 
  ! Step 9: Clean up ELPA

  call elpa_deallocate(elpaInstance, status)
  ! Check status code ...
  call elpa_uninit()

  ! ____________________________________________ 
  ! Clean up the rest and finalize
  deallocate(A_loc) 
  deallocate(Z_loc)
  deallocate(Eigenvalues)
  
  call blacs_gridexit(ictxt)
  call MPI_Finalize(ierr)
  
contains

! Function to compute the global matrix element of "Symmetrized_Clement" matrix
function get_global_matrix_element(I_gl, J_gl, n) result(matrix_element)
  implicit none
  integer, intent(in) :: I_gl, J_gl, n
  real(8) :: matrix_element
  integer :: k

  matrix_element = 0.0d0
  if (I_gl == J_gl+1 .or. J_gl == I_gl+1) then
    K = min(I_gl, J_gl)
    matrix_element = dsqrt(dble(k * (n - k)))
  end if

end function get_global_matrix_element

end program